source/common/upstream/subset_lb.cc

```diff
@@ -101,6 +106,70 @@ SubsetLoadBalancer::~SubsetLoadBalancer() {
   });
 }
 
+void SubsetLoadBalancer::rebuildSingle() {
+  if (single_key_.empty()) {
+    return;
+  }
+
+  // Because PriorityUpdateCb doesn't give a modified list (only added and removed), it is
+  // faster to just rebuild this map than try to figure out if any hosts had their metadata
+  // changed, and then figure out the old and new value (to remove from the old key in this map
+  // and insert in the new key).
+  single_host_per_subset_map_.clear();
+
+  uint32_t collision_count = 0;
+  for (const auto& host_set : original_priority_set_.hostSetsPerPriority()) {
+    for (const auto& host : host_set->hosts()) {
+      MetadataConstSharedPtr metadata = host->metadata();
+      const auto& filter_metadata = metadata->filter_metadata();
+      auto filter_it = filter_metadata.find(Config::MetadataFilters::get().ENVOY_LB);
+      if (filter_it != filter_metadata.end()) {
+        const auto& fields = filter_it->second.fields();
+        auto fields_it = fields.find(single_key_);
+        if (fields_it != fields.end()) {
+          auto [iterator, did_insert] =
+              single_host_per_subset_map_.try_emplace(fields_it->second, host);
+          if (!did_insert) {
+            // Two hosts with the same metadata value were found. Ignore all but one of them, and
+            // set a metric for how many times this happened.
+            collision_count++;
+          }
+        }
+      }
+    }
+  }
+
+  // This stat isn't added to `ClusterStats` because it wouldn't be used
+  // for nearly all clusters, and is only set during configuration updates,
+  // not in the data path, so performance of looking up the stat isn't critical.
+  Stats::Utility::gaugeFromElements(
+      scope_, {Stats::DynamicName("lb_subsets_single_host_per_subset_duplicate")},
+      Stats::Gauge::ImportMode::Accumulate)
+      .set(collision_count);
+}
+
+// When in `single_host_per_subset` mode, select a host based on the provided match_criteria.
+// Set `host_chosen` to
+HostConstSharedPtr SubsetLoadBalancer::tryChooseHostFromMetadataMatchCriteriaSingle(
+    const Router::MetadataMatchCriteria& match_criteria, bool& host_chosen) {
+  ASSERT(!single_key_.empty());
+
+  for (const auto& entry : match_criteria.metadataMatchCriteria()) {
+    if (entry->name() == single_key_) {
+      auto it = single_host_per_subset_map_.find(entry->value());
+      if (it != single_host_per_subset_map_.end()) {
+        if (it->second->health() != Host::Health::Unhealthy) {
```

Comment:
I can imagine wanting this to be configurable, I can see wanting to try to connect to the host regardless of health. Probably fine to leave as is, can add later if requested